### ISSUE TITLE
Assure consistent and expected behavior in staradmin start/stop db/host

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Home#130]
Currently, `staradmin` support management of database related processes via

* staradmin start db
* staradmin stop db
* staradmin stop host

The `stop db` command will assure both any host, as well as the database process(es), are stopped, while `stop host` will stop just the host, but leave database process(es) left untouched. Probably least surprise principle applied here, and quite consistent.

As for `start db`, it will start both database process(es) and the shared host, but **only start the host if the database process(es) are not running**. Hence:

```
staradmin start server
staradmin start db // will start database + host
staradmin stop host // will stop host, but leave database
staradmin start db // the second call will not start the host - inconsistent
```

1. We'd like second invocation of `start db` to also start shared host, right?
2. It's currently not possible to start **only** database processes via CLI (i.e. those, but not the host), unless you use combination of `staradmin start db` + `staradmin stop host`. Is this something we want to introduce? If so, we need to review the whole naming I think.

(Also: this will likely change over time, in 2.4 / 3.x / Core, where it's not even sure if explicit termination of database processes will be allowed. But I think at least we need to fix [1], and possibly already in 2.3).

**Comment by k-rus** [comment]
We need a way to start host if it is not started.

**Comment by per-samuelsson** [comment]
> We need a way to start host if it is not started.

`staradmin start db` would do that today already (and better so after fixing [1] from above list). I'm thinking explicitly supporting `staradmin start host` would make no sense, since the host can not run without the database.

**Comment by k-rus** [comment]
> We need a way to start host if it is not started.
> staradmin start db would do that today already (and better so after fixing [1] from above list). I'm thinking explicitly supporting staradmin start host would make no sense, since the host can not run without the database.

Thanks for clarifying!

**Comment by per-samuelsson** [comment]
One idea is we keep, but *officially* stop supporting, `staradmin start db`.

My thinking is both `stop app`, `stop db` and `stop host` (i.e. "stop all apps") fill a real purpose, while `start db` does not. Instead, we make starting always implicit - with both backend processes and shared host only starting on demand, when users run their apps.

We keep `start db` option supported for some time (at least in 2.4) and hence we are still backwards compatible, and we can use it for our purposes - i.e. when we want to start the shared host to debug it before apps get loaded. But remove it from our docs and help.

WDYT?

**Comment by k-rus** [comment]
> My thinking is both stop app, stop db and stop host (i.e. "stop all apps") fill a real purpose, while start db does not. Instead, we make starting always implicit - with both backend processes and shared host only starting on demand, when users run their apps.
> 
> We keep start db option supported for some time (at least in 2.4) and hence we are still backwards compatible, and we can use it for our purposes - i.e. when we want to start the shared host to debug it before apps get loaded. But remove it from our docs and help.

For the debugging purpose we and our customers, I believe, extremely need the ability to start database and code host without any application. So on demand doesn't cover our need.

**Comment by per-samuelsson** [comment]
> For the debugging purpose we and our customers, I believe, extremely need the ability to start database and code host without any application.

Well, as I wrote, we keep it around, because I also see a need **we** need it for the purpose of debugging, which is why I wrote:

> We keep `start db` option supported for some time (at least in 2.4) and hence we are still backwards compatible, **and we can use it for our purposes - i.e. when we want to start the shared host to debug it before apps get loaded**.

But in what way do you think customers need it. Can you give a concrete case?

Remember, trimming our surface area and supporting less options is for the *good*. It makes our product easier to understand and use. I think that when we find exposures that are not used by our customers, we should really strive to hide them away.

**Comment by k-rus** [comment]
> But in what way do you think customers need it. Can you give a concrete case?

Customer wants to debug an application at the Main entry point. I guess it is not a problem if the application is created as Starcounter project through the Starcounter template. However, it is not obvious how to do if it is not a Starcounter project, e.g., it is plain C# file (not a project), which is totally valid, or unsupported version of VS. Then I will do it by starting the code host and attaching to `sccode` from VS and then starting the application.

**Comment by per-samuelsson** [comment]
> However, it is not obvious how to do if it is not a Starcounter project, e.g., it is plain C# file (not a project), which is totally valid, or unsupported version of VS. Then I will do it by starting the code host and attaching to `sccode` from VS and then starting the application.

Thanks. That's a valid case I did not consider.

Still, that use-case is pretty slim I think and alternatives exist:

1. We will keep `staradmin start db` for quite some time, just undocumented.
2. `Debugger.Launch` in entrypoint would work.
3. Creating another "dummy" app would work, and launching that first.
4. Turning it into a real project is always a valid option, and might be good if you want to regularly debug the entrypoint.

On the plus side of deprecating `staradmin start db` is that our tooling become more uniform - Starcounter becomes a more clear platform for running apps. You can stop apps, or even stop them in batch (i.e. the host), using `staradmin´ but to **start** anything, you always use `star`, your (self-hosted) exe or web admin.

**Comment by k-rus** [comment]
> We will keep staradmin start db for quite some time, just undocumented.

I am worry about keeping only for some time.


> Debugger.Launch in entrypoint would work.

It will pollute the actual code. It requires modifying the code and then recompiling it.


> Creating another "dummy" app would work, and launching that first.

This approach is already in use for quite some time and there were advocates against it. I think it is not a nice approach.



> Turning it into a real project is always a valid option, and might be good if you want to regularly debug the entrypoint.

This doesn't work for customers using unsupported versions of VS such as 2012 and 2013.

My point is that it would be nice to be able to start the code host process without a dummy project. I don't have strong opinion about actual interface to user for allowing this. I guess it will stop work in some future when multi-hosts will be supported, but the need will be still there.

**Comment by per-samuelsson** [comment]
Well, ok, let's keep it, and fix so that it's fully consistent and well documented. Thanks for input.

**Comment by per-samuelsson** [comment]
I'm treating [1] from OP listing as a bug, and will fix this inconsistency already in 2.3.

**Comment by per-samuelsson** [comment]
Fixed. Merged and pushed in https://github.com/Starcounter/level1/commit/e32b31f64405af7f316ac606f5f2316b61743186, and then merged to develop.
